Small refactor of the RGBDSensorClient device

drdanz opened the pull request:
## Bug Fixes

### Devices

#### `RGBDSensorClient`

* The device no longer crashes when no image was received yet (#2349).
* The timestamp is now correct.

----

Fixes #2349

PeterBowman commented on src/devices/RGBDSensorClient/RGBDSensorClient_StreamingMsgParser.cpp:
```diff
@@ -9,67 +9,97 @@
+void RgbImageBufferedPort::onRead(yarp::sig::FlexImage& datum)
 {
-    last_rgb = datum;
+    local_arrival_time = yarp::os::Time::now();
+    std::lock_guard<std::mutex> lock(mutex);
```
The lock is acquired after modifying `local_arrival_time`. Is this intentional? Note that in `getImage` this variable is accessed within the critical section, hence my doubt.

PeterBowman commented on src/devices/RGBDSensorClient/RGBDSensorClient_StreamingMsgParser.cpp:
```diff
@@ -9,67 +9,97 @@
+void FloatImageBufferedPort::onRead(yarp::sig::ImageOf< yarp::sig::PixelFloat> & datum)
 {
-    last_depth = datum;
+    local_arrival_time = yarp::os::Time::now();
+    std::lock_guard<std::mutex> lock(mutex);
```
Same here (first `local_arrival_time`, then mutex lock).

PeterBowman commented on src/devices/RGBDSensorClient/RGBDSensorClient_StreamingMsgParser.cpp:
```diff
@@ -9,67 +9,97 @@
+        // No image received yet
+        return std::make_tuple(false, yarp::sig::ImageOf<yarp::sig::PixelFloat>(), yarp::os::Stamp());
+    }
+    return std::make_tuple(true, last_depth, stamp);
```
I believe here (and in other places like this) you should be allowed to reduce it to:

```c++
return {true, last_depth, stamp};
```

PeterBowman replied:
You might want to disregard this comment... There is something fishy around `std::tuple`, look for "conditionally-explicit constructors" in https://en.cppreference.com/w/cpp/utility/tuple/tuple. According to https://stackoverflow.com/a/16713413, this was a compiler extension back in 2013 or so, but later (partially?) fixed in C++11 via [N4387](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4387.html).

drdanz replied:
Yes, unfortunately N4387 is in C++17 and we still require c++14.
It's probably worth add a FIXME comment to remember changing it when we have the compiler support

PeterBowman reviewed (commented):
Suggestion: `RGBDSensor_StreamingMsgParser::read` is not called anywhere, but it could be of use in `RGBDSensorClient::getImages` as opposed to calling both `readRgb` and `readDepth` separately.

Nicogene reviewed (commented):
The target branch should be `yarp-3.4` ? Since it is a bugifix

drdanz commented:
> The target branch should be `yarp-3.4` ? Since it is a bugifix

This device was modified on `master`, and it would cause conflicts, therefore I'm pushing this commit to master.

drdanz commented:
Fixed all suggestions (thanks @PeterBowman)

> Suggestion: `RGBDSensor_StreamingMsgParser::read` is not called anywhere, but it could be of use in `RGBDSensorClient::getImages` as opposed to calling both `readRgb` and `readDepth` separately.

While fixing this, I noticed that the timestamp returned was still generated by the client, therefore I fixed it as well.

If you are interested in having this in the YARP 3.4 branch, I can try to cherry pick the commits after merging this

PeterBowman commented:
For me it's perfectly fine to leave these commits in master, many thanks for fixing this :).
